// @runette/ngx-leaflet-locate/esm2015/lib/ngx-leaflet-locate.component.js
/// <reference path="../../node_modules/@types/leaflet.locatecontrol/index.d.ts" />
/// <reference path="../../node_modules/@types/leaflet.locatecontrol/index.d.ts" />
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { control, Control } from 'leaflet';
import '../../../../node_modules/leaflet.locatecontrol/dist/L.Control.Locate.min.js';
import * as i0 from "@angular/core";
export class NgxLeafletLocateComponent {
    constructor() {
        this.location$ = new EventEmitter;
        this.control = new Control.Locate();
        this.options = {};
    }
    ;
    ngOnInit() {
    }
    ;
    ngOnDestroy() {
        var _a, _b, _c;
        if (this.control && this.map) {
            this.control.stop();
            (_a = this._map) === null || _a === void 0 ? void 0 : _a.off("unload", (_b = this.map) === null || _b === void 0 ? void 0 : _b._unload, this.control);
        }
        if (this._map && this.control)
            this._map.removeControl(this.control);
        (_c = this._map) === null || _c === void 0 ? void 0 : _c.off('locationfound');
    }
    ;
    set map(map) {
        if (map) {
            this._map = map;
            this.control = control.locate(this.options);
            this.control.addTo(map);
            let location$ = this.location$;
            map.on('locationfound', e => {
                location$.emit(e);
            });
        }
    }
    get map() {
        return this._map;
    }
}
NgxLeafletLocateComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: NgxLeafletLocateComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NgxLeafletLocateComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: NgxLeafletLocateComponent, selector: "leaflet-locate-control", inputs: { options: "options", map: "map" }, outputs: { location$: "location$" }, ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: NgxLeafletLocateComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'leaflet-locate-control',
                    template: '',
                    styleUrls: []
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { location$: [{
                type: Output
            }], options: [{
                type: Input
            }], map: [{
                type: Input
            }] } });
