// @runette/ngx-leaflet-locate/esm2015/lib/ngx-leaflet-locate.module.js
import { NgModule } from '@angular/core';
import { NgxLeafletLocateComponent } from './ngx-leaflet-locate.component';
import * as i0 from "@angular/core";
export class NgxLeafletLocateModule {
}
NgxLeafletLocateModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: NgxLeafletLocateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxLeafletLocateModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: NgxLeafletLocateModule, declarations: [NgxLeafletLocateComponent], exports: [NgxLeafletLocateComponent] });
NgxLeafletLocateModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: NgxLeafletLocateModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: NgxLeafletLocateModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxLeafletLocateComponent],
                    imports: [],
                    exports: [
                        NgxLeafletLocateComponent,
                    ]
                }]
        }] });
