// @runette/ngx-leaflet-fullscreen/esm2020/lib/ngx-leaflet-fullscreen.module.mjs
import { NgModule } from '@angular/core';
import { FullscreenControlComponent } from './ngx-leaflet-fullscreen.component';
import * as i0 from "@angular/core";
export class NgxLeafletFullscreenModule {
}
NgxLeafletFullscreenModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NgxLeafletFullscreenModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxLeafletFullscreenModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: NgxLeafletFullscreenModule, declarations: [FullscreenControlComponent], exports: [FullscreenControlComponent] });
NgxLeafletFullscreenModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NgxLeafletFullscreenModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NgxLeafletFullscreenModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FullscreenControlComponent],
                    imports: [],
                    exports: [FullscreenControlComponent]
                }]
        }] });
