// @runette/ngx-leaflet-fullscreen/esm2020/lib/ngx-leaflet-fullscreen.component.mjs
/// <reference types='@runette/leaflet-fullscreen'/>
/// <reference types='@runette/leaflet-fullscreen'/>
import { Component, Input } from '@angular/core';
import { Control } from 'leaflet';
import '../../../../node_modules/@runette/leaflet-fullscreen/dist/Leaflet.fullscreen.min.js';
import * as i0 from "@angular/core";
export class FullscreenControlComponent {
    constructor() {
        this._options = {};
    }
    set options(opt) {
        this._options = opt;
        this.control = new Control.Fullscreen(this._options);
        if (this._map) {
            this._map?.addControl(this.control);
            this._map?.on('enterFullscreen', () => this._map?.invalidateSize());
            this._map?.on('exitFullscreen', () => this._map?.invalidateSize());
        }
    }
    ngOnInit() {
    }
    ngOnDestroy() {
        if (this.control) {
            this._map?.removeControl(this.control);
            this._map?.off('enterFullscreen');
            this._map?.off('exitFullscreen');
        }
    }
    set map(map) {
        if (map) {
            this._map = map;
            if (this.control) {
                map.addControl(this.control);
                map.on('enterFullscreen', () => map.invalidateSize());
                map.on('exitFullscreen', () => map.invalidateSize());
            }
        }
        ;
    }
    ;
    get map() {
        return this._map;
    }
    ;
}
FullscreenControlComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: FullscreenControlComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FullscreenControlComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: FullscreenControlComponent, selector: "leaflet-fullscreen-control", inputs: { options: "options", map: "map" }, ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: FullscreenControlComponent, decorators: [{
            type: Component,
            args: [{ selector: 'leaflet-fullscreen-control', template: '' }]
        }], ctorParameters: function () { return []; }, propDecorators: { options: [{
                type: Input
            }], map: [{
                type: Input
            }] } });
;
