// @runette/ngx-leaflet-loading/esm2020/lib/ngx-loading-control.component.mjs
/// <reference types='leaflet-loading' />
/// <reference types='leaflet-loading' />
import { Component, Input } from '@angular/core';
import { Control } from 'leaflet';
import '../../../../node_modules/leaflet-loading/src/Control.Loading.js';
import * as i0 from "@angular/core";
export class NgxLoadingControlComponent {
    constructor() {
        this.loading = new Control();
        this.options = {};
    }
    ngOnInit() {
    }
    ngOnDestroy() {
        this._map?.removeControl(this.loading);
    }
    set map(map) {
        if (map) {
            this._map = map;
            this.loading = new Control.Loading(this.options);
            this.loading.addTo(map);
        }
    }
    get map() {
        return this._map;
    }
}
NgxLoadingControlComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: NgxLoadingControlComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NgxLoadingControlComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.8", type: NgxLoadingControlComponent, selector: "leaflet-loading-control", inputs: { options: "options", map: "map" }, ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: NgxLoadingControlComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'leaflet-loading-control',
                    template: '',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { options: [{
                type: Input
            }], map: [{
                type: Input
            }] } });
