// @runette/ngx-leaflet-loading/esm2020/lib/ngx-loading-control.module.mjs
import { NgModule } from '@angular/core';
import { NgxLoadingControlComponent } from './ngx-loading-control.component';
import * as i0 from "@angular/core";
export class NgxLoadingControlModule {
}
NgxLoadingControlModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: NgxLoadingControlModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxLoadingControlModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: NgxLoadingControlModule, declarations: [NgxLoadingControlComponent], exports: [NgxLoadingControlComponent] });
NgxLoadingControlModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: NgxLoadingControlModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: NgxLoadingControlModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxLoadingControlComponent],
                    imports: [],
                    exports: [NgxLoadingControlComponent]
                }]
        }] });
